Engines and Node 18+Windows tests

lukekarrys opened the pull request:
BREAKING CHANGE: `npm` is now compatible with the following semver range
for node: `^14.17.0 || ^16.0.0 || >=18.0.0`

Ref: https://github.com/npm/statusboard/issues/519

lukekarrys commented:
Engines updates to workspaces will be updated by template-oss once https://github.com/npm/template-oss/pull/180 lands.

ljharb commented:
Unfortunately nvm's CI on travis-ci is currently broken (because they changed a docker container out from under me) so i won't be able to cut a release until that's fixed, so `nvm install-latest-npm` will break for users on node `^12.13` (which includes all of my packages).

I don't think this is npm's problem to solve, to be clear, just giving a heads up.

ljharb commented on package-lock.json:
```diff
@@ -178,7 +178,7 @@
       },
       "engines": {
-        "node": "^12.13.0 || ^14.15.0 || >=16"
+        "node": "^14.17.0 || ^16.0.0 || >=18.0.0"
```
why 14.17 and not 14.15, since 14.15 is when it entered LTS?

also why 16.0.0 and not v14.13.0 when it entered LTS?

lukekarrys replied:
Mostly as an attempt to align with other packages we consume in the ecosystem. I wanted to set our engines as low as we could, while still allowing for packages like `readable-stream@4` to be installed (which has its engines field set to [`^12.22.0 || ^14.17.0 || >=16.0.0`](https://github.com/nodejs/readable-stream/blob/77e0378b1a8a46c42fc3fce43969b83abf69ca36/package.json)).

Same logic applies to 16. I didn't see any reason to raise this floor since we wont be able to consume any packages that changed to `> 16.0.0` until `npm10` when this will all get revisited.

lukekarrys replied:
This is a summary of https://github.com/npm/statusboard/issues/519, which goes through the other engines values we discussed.

lukekarrys replied:
After typing this out I now realize that a dependency could ship with `^14.17.0 || ^16.13.0 || >=18` and we couldn't be able to consume it so maybe there is a case for raising the floor for 16.

lukekarrys replied:
The engines is now `^14.17.0 || ^16.13.0 || >=18`

lukekarrys commented:
> Unfortunately nvm's CI on travis-ci is currently broken (because they changed a docker container out from under me) so i won't be able to cut a release until that's fixed, so `nvm install-latest-npm` will break for users on node `^12.13` (which includes all of my packages).

This will be released as `9.0.0-pre.0` and pushed to the `next-9` dist-tag initially and won't go out as `9.0.0` and `latest` for a little bit.  Does that affect how `install-latest-npm` will work?

ljharb commented:
It won't break anything until it's published as "latest".

wraithgar commented:
The thing that bit us in the past was when `node-gyp` added an engines declaration that was more restrictive than ours.  We'll be more proactive about this in the future, making sure we speak up if that's about to happen again there, and we will just have to take other third party engines updates as they come.  It doesn't appear many packages right now are adopting things in 16 that require them to set an engines more restrictive than `>=16` so leaving this PR as-is will likely work.

The `Pin to LTS` decision of `npm@8` was one born out of "we have to come up with something and this seems good". We know more now than we did then, including the fact that node backports new features in LTS lines after a version has gone into LTS.  Pinning to LTS as the floor doesn't help us as much as we thought it did.

I think this PR should be merged as-is.  We will likely be revisiting the `engines` field more often in the future than in the past so it won't be years before we can correct something.

ljharb commented:
Major versions are always hugely painful for the ecosystem and cause lots of churn. If a less restrictive engines range means there will be more major versions, then I implore you to make it as restrictive as possible, with the goal of having as few major bumps as possible.

lukekarrys committed:
chore: fix failing tests in windows + node 18

Some of our tests were failing in windows after testing on node 18. The
reason was the inability to clean up the logs dir. This changes forces
a few tests to run in order and also cleans up any use of multiple
`t.testdir` calls in a single child test which can cause problems.

wraithgar reviewed (changes requested):
Second commit should be a  `fix` and include info about what changed in `lib/cli.js`

wraithgar committed:
feat: update supported node engines in package.json

This also replaces the previous check for known broken versions of node
with an exception handler for syntax errors in order to try and give a
nicer error message when attempting to run npm on older node versions.

BREAKING CHANGE: `npm` is now compatible with the following semver range
for node: `^14.17.0 || ^16.13.0 || >=18.0.0`

Ref: npm/statusboard#519

wraithgar committed:
fix: attempt more graceful failure in older node versions

wraithgar commented:
Pulled the cli.js changes into a separate `fix`

wraithgar reviewed (approved):
Waiting on CI, everything looks good now.
